File: app/src/main/java/com/heptad/app/data/models/Puzzle.kt
package com.heptad.app.data.models

import java.util.UUID

/**
 * Represents a Heptad puzzle configuration.
 *
 * @property id Unique identifier for the puzzle
 * @property centerLetter The mandatory center letter (must be included in all valid words)
 * @property outerLetters The 6 outer/orbital letters
 * @property validWords All valid words that can be formed with this puzzle
 * @property pangrams Words that use all 7 letters at least once
 * @property maxScore Maximum achievable score for this puzzle
 * @property createdAt Timestamp when the puzzle was created
 * @property dictionaryLevel The SCOWL dictionary level used (60, 70, or 80)
 * @property includesS Whether the letter 'S' was allowed during generation
 */
data class Puzzle(
    val id: String = UUID.randomUUID().toString(),
    val centerLetter: Char,
    val outerLetters: List<Char>,
    val validWords: Set<String>,
    val pangrams: Set<String>,
    val maxScore: Int,
    val createdAt: Long = System.currentTimeMillis(),
    val dictionaryLevel: Int = 70,
    val includesS: Boolean = true
) {
    /**
     * All 7 letters in the puzzle (center + outer)
     */
    val allLetters: Set<Char>
        get() = (outerLetters + centerLetter).toSet()

    /**
     * Total number of valid words
     */
    val wordCount: Int
        get() = validWords.size

    /**
     * Check if a word is a pangram
     */
    fun isPangram(word: String): Boolean = word.lowercase() in pangrams

    /**
     * Check if a word is valid for this puzzle
     */
    fun isValidWord(word: String): Boolean = word.lowercase() in validWords

    /**
     * Calculate score for a word
     * - 4-letter words: 1 point
     * - 5+ letter words: 1 point per letter
     * - Pangram bonus: +7 points
     */
    fun calculateWordScore(word: String): Int {
        val normalizedWord = word.lowercase()
        if (normalizedWord !in validWords) return 0

        val baseScore = if (normalizedWord.length == 4) 1 else normalizedWord.length
        val pangramBonus = if (normalizedWord in pangrams) 7 else 0
        return baseScore + pangramBonus
    }

    companion object {
        /**
         * Create an empty/placeholder puzzle
         */
        fun empty(): Puzzle = Puzzle(
            id = "empty",
            centerLetter = 'A',
            outerLetters = listOf('B', 'C', 'D', 'E', 'F', 'G'),
            validWords = emptySet(),
            pangrams = emptySet(),
            maxScore = 0
        )
    }
}

File: web/src/lib/models/puzzle.ts
export interface Puzzle {
  id: string;
  centerLetter: string;
  outerLetters: string[];
  validWords: Set<string>;
  pangrams: Set<string>;
  maxScore: number;
  dictionaryLevel: number;
  includesS: boolean;
}

export function createPuzzle(params: {
  id?: string;
  centerLetter: string;
  outerLetters: string[];
  validWords: string[] | Set<string>;
  pangrams: string[] | Set<string>;
  maxScore: number;
  dictionaryLevel?: number;
  includesS?: boolean;
}): Puzzle {
  return {
    id: params.id ?? crypto.randomUUID(),
    centerLetter: params.centerLetter.toLowerCase(),
    outerLetters: params.outerLetters.map(l => l.toLowerCase()),
    validWords: params.validWords instanceof Set ? params.validWords : new Set(params.validWords.map(w => w.toLowerCase())),
    pangrams: params.pangrams instanceof Set ? params.pangrams : new Set(params.pangrams.map(w => w.toLowerCase())),
    maxScore: params.maxScore,
    dictionaryLevel: params.dictionaryLevel ?? 70,
    includesS: params.includesS ?? true
  };
}

export function getAllLetters(puzzle: Puzzle): Set<string> {
  return new Set([...puzzle.outerLetters, puzzle.centerLetter]);
}

export function isPangram(puzzle: Puzzle, word: string): boolean {
  return puzzle.pangrams.has(word.toLowerCase());
}

export function calculateWordScore(puzzle: Puzzle, word: string): number {
  const w = word.toLowerCase();
  if (!puzzle.validWords.has(w)) return 0;
  const baseScore = w.length === 4 ? 1 : w.length;
  const pangramBonus = puzzle.pangrams.has(w) ? 7 : 0;
  return baseScore + pangramBonus;
}
